### DISCUSSION TITLE
🎧 Kaizen 12 - Slightly more instant - Oct. 13, 2023

### DISCUSSION THREAD

**Discussion by gerhard** [thechangelog/changelog.com#480] [Kaizen]
[![shields.io](https://img.shields.io/badge/Last%20updated%20on%20October%2021%2C%202023-PUBLISHED-blue?style=for-the-badge)](https://changelog.com/friends/17)

[🎧 This is now available as changelog.com/friends/17](https://changelog.com/friends/17)

<a href="https://changelog.com/friends/17" target="_blank"><img width="1440" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/613ef497-af96-409c-a358-63e8c1315e13"></a>

See all improvements in this 🐙 GitHub repository: [Kaizen 11 ... 12](https://github.com/thechangelog/changelog.com/compare/kaizen.11...kaizen.12)

---

This contains all things related to the next Kaizen episode.

Share your thoughts via comments below:
- Is there something in changelog.com that you, our dear listener, would like us to tackle?
- Is there something that **you** would like to help out with?
- Anything that we - @adamstac @jerodsanto - want to improve?

A rough timeline of what to expect (red line is today):

```mermaid
gantt
    title 🎧 Kaizen 12 timeline
    dateFormat  YYYY-MM-DD
    axisFormat %b %e, %Y

    section DISCUSS
    ✍️ Add your comments below :2023-08-11, 61d

    section CONTRIBUTE
    💻 ALL CONTRIBUTIONS WELCOME 👕 :2023-08-11, 61d

    section RECORD


    🎙 Record :milestone, 2023-10-11
    ✂️ Edit :2023-10-11, 3d
    🚀 SHIP IT! :milestone, 2023-10-14

    section NEXT
    Kick-off Kaizen 13 :2023-10-14, 14d
```

---

#### What we **didn't** do: multi-node clustering & caching
- Did procrastinating pay off?

#### Serving all our feeds from R2, via Fastly
- How did that go Jerod?
- Pairing sessions with James A. Rosen - @jamesarosen
  - We were holding AWS S3 & Fastly integration wrong - config now removed from Fastly)
- Cached `/feed` tail latency improved by `4x`, and cache hit ratio improved by `3x`
- See the GitHub discussion for more details
  - Fastly debugging tips
  - Analyzing HTTP response headers
  - Honeycomb.io before & after screenshots
  - James's `CacheHitRatio` formula of type `math` Markdown
- SLO [`96%` of all feeds serve within `1s` - last `60 days`](https://ui.honeycomb.io/changelog/datasets/fastly/slo/3vSeCSXS671/Podcast-feed-response-latency) - notice the gain since Aug. 31, 2023

#### Oban Web - PR #472 
- _Well I finally got Fastly websockets passthrough configured correctly and now Oban Web is working like expected_
  - Your free trial of WebSockets ends on October 29, 2023
- Big shout-out to Parker Selbert - @sorentwo
- Check the debugging chat in our Slack `#dev` channel - don't archive it!
- Small shout-out to Lars Wikman - @lawik 
- See PR #472 for all the details, including how to integrate it in a Dagger CI/CD pipeline

#### Migrating nightly.changelog.com from DigitalOcean to Fly.io
- https://github.com/thechangelog/nightly/pull/42
  - Containerize & Deploy, a.k.a. `daggerize`
- Cloudflare Pages experiment
  - `wrangler pages project list`
  - <https://nightly.changelog.place>
- Cloudflare R2 experiment - `index.html`
- Resolves https://github.com/thechangelog/changelog.com/issues/420

#### Honeycomb.io
-  SLOs vs Triggers (see the GitHub discussion for more details)
  - [Elevated 5xx HTTP responses](https://ui.honeycomb.io/changelog/datasets/fastly/triggers/5U7Huz7uFfa) trigger - `15mins`
  - [Elevated 404 HTTP responses](https://ui.honeycomb.io/changelog/datasets/fastly/triggers/8pENCV13L76) trigger - `60mins`
- SLO [`99%` of all responses are `2xx` or `3xx` - last `60 days`](https://ui.honeycomb.io/changelog/datasets/fastly/slo/4mzdutPaJvf/2xx-or-3xx-responses) - handle some of those 404s
  - [Reduce 404 responses](https://ui.honeycomb.io/changelog/datasets/fastly/result/kVJ9kZomjEW) - [`1.46%` of all traffic](https://manage.fastly.com/observe/dashboard/system/overview/historic/7gKbcKSKGDyqU7IuDr43eG?range=30day)

#### Fastly
- Enable HTTP/3 QUIC
  - Noticed the absence of HTTP/3 responses in [Fastly's beta Edge observer](https://manage.fastly.com/observe/dashboard/system/overview/historic/7gKbcKSKGDyqU7IuDr43eG?range=1hr)
- Use Brotli compression when available
  - 10-15% reduction in traffic over gzip - How can we measure that?
- [Now redirecting `/rss` at the edge](https://ui.honeycomb.io/changelog/datasets/fastly/result/vkFHdoq1hWJ?hideCompare) - `400ms` vs `1ms`

#### KubeCon NA 2023
- Meeting for the first time in person - 7 years in the making!
- Who do we want to record with?
  - Frederic from Polar Signals - recently launched [PolarSignals Cloud](https://cloud.polarsignals.com/)
    - [Ship It! #57](https://changelog.com/shipit/57) was the last time Gerhard recorded with him
    - Find previous [conversations with Frederic on Changelog](https://changelog.com/person/fredbrancz) - starting with KubeCon 2019!
    - Our Fly.io experiment with PolarSignals Cloud: https://www.polarsignals.com/docs/setup-collection-kubernetes/
  - Erik, a BuildKit maintainer, that has been doing some really cool things with distributed caching
    - S3 R2 B2 on steroids
  - Solomon!
- Dagger @ KubeCon!
  - Come say hi to Gerhard at booth `N37`

---

### What changed so far? [Kaizen.11 ... NOW](https://github.com/thechangelog/changelog.com/compare/kaizen.11...master)
- #482 
- #472
- #451
- https://github.com/thechangelog/nightly/pull/42

---

### What else is coming up?
- Parca on Fly.io -> PolarSignals Cloud - https://www.polarsignals.com/docs/setup-collection-kubernetes/
- [Reduce 404s](https://ui.honeycomb.io/changelog/datasets/fastly/result/xaa4Q7U4Aaq)
- Experiment with [middleware.io](https://middleware.io/)
- Elixir / Phoenix releases
- 1Password Service Accounts
- https://github.com/thechangelog/changelog.com/issues/419

---

### This is a follow-up to
- https://github.com/thechangelog/changelog.com/discussions/469
- https://github.com/thechangelog/changelog.com/discussions/452
- https://github.com/thechangelog/changelog.com/discussions/440
- https://github.com/thechangelog/changelog.com/discussions/433

**Comment by gerhard** [comment]
This is a placeholder thread for talking fastly.com integration async with @jamesarosen: https://changelog.slack.com/archives/C03SA8VE2/p1692485028316819

<img width="998" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/cbd97723-2419-4857-9d92-0ba6e25ee150">

**Reply by gerhard** [reply]
## Changelog & R2

As a first step, we ensured that we are returning the correct header to Fastly so that static assets coming from R2 will get cached correctly. This is the end-result:

<img width="1436" alt="Screenshot 2023-09-01 at 19 53 14" src="https://github.com/thechangelog/changelog.com/assets/3342/5969504c-f9ab-4d9e-8f1b-43aac2ea496c">

The [`cache-control` header that we set in Cloudflare for all `changelog.place` responses](https://dash.cloudflare.com/acd4d0fe190cbd98417069601607c33a/changelog.place/rules/transform-rules/modify-response-header/d5a00b1ea6f54a74b31d151063bd8e9f) is saying:
- `max-age=3600` - clients should cache for `1h`
- `s-maxage=604800` - a proxy cache (e.g. Fastly) should cache for `7d`
- `stale-while-revalidate=604800` - cache can serve a stale response for another `7d` while it fetches the object from the origin, typically in the background
- `stale-if-error=604800` - cache can serve a stale response for `7d` if the origin returns an error (usually a `5xx` response)

The other 2 headers in the above screenshot are:
- `cf-cache-status: HIT` - this response was served from the Cloudflare cache
- `age: 36` - it was stored in the Cloudflare cache for `36s` ago at the time the response was served

FTR:
<img width="1445" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/1a253e44-89dd-43b8-abf6-9c7761088532">

## Fastly

Now let's see what Fastly is returning for the same url:
<img width="1433" alt="Screenshot 2023-09-01 at 20 11 58" src="https://github.com/thechangelog/changelog.com/assets/3342/d8b9a70e-12ee-4f92-94e6-2f40536406db">

We see that the original `cache-control` headers set in Cloudflare are preserved.

We also see that the request was not cached in the Cloudflare POP that serviced the Fastly request.

The headers highlighted in cyan are most interesting:
- `x-served-by: cache-iad-kjyo7100132-IAD, cache-lon420136-LON` - request was served by the Fastly `LON` edge POP (closest to my location), which requested it from the Fastly `IAD` POP (our cache shield)
- `x-cache: MISS, HIT` - request was served from the edge cache (the second `HIT`), and it was not in the shield cache at the time

> **Note**
> This last part is really important. While this might look like a miss, meaning that it was not served from the Fastly cache, only the initial request to the cache shield - `IAD` - was a miss. If the edge cache has it - `LON` -, then the cache shield loaded it too when the edge requested it.

Adding `Fastly-Debug: 1` header shows us more:
<img width="1434" alt="Screenshot 2023-09-01 at 20 33 17" src="https://github.com/thechangelog/changelog.com/assets/3342/2a7e1aa4-c4ec-482c-983d-693a30ccbbaf">

[This Fastly.com doc page](https://docs.fastly.com/en/guides/checking-cache) helps with understanding the symbols.

## mp3 responses from cdn.changelog since R2

`WIP`

<img width="1183" alt="Screenshot 2023-09-01 at 20 36 23" src="https://github.com/thechangelog/changelog.com/assets/3342/ed6f7b06-9079-4f98-b97d-a476f8aa2bb0">

## AWS S3

`WIP`

<img width="1435" alt="Screenshot 2023-09-01 at 20 38 00" src="https://github.com/thechangelog/changelog.com/assets/3342/82321359-d2a3-4f72-aeda-f07779790878">

---

Links:
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control
- https://docs.fastly.com/en/guides/checking-cache
- https://developers.cloudflare.com/cache/how-to/cache-rules/

**Reply by jamesarosen** [reply]
> `x-cache: MISS, HIT` - request was served from the edge cache (the second `HIT`)

Correct.

> and it was not in the shield cache at the time

This is either correct or incorrect depending on the reading of the sentence. It was not in Shield cache _at the time Edge first requested it from Shield and then placed it into its cache_. It may or may not be in Shield at the point when Edge is serving a HIT, but Edge doesn't know because the HIT means it doesn't bother connecting to Shield.

**Reply by jamesarosen** [reply]
### Cache-Hit Ratio

If you want to know how much work Fastly (Edge+Shield combined) is offloading from R2, you want

```math
CHR = \frac{(HIT,HIT) + (HIT,MISS) + (MISS,HIT)}{(HIT,HIT) + (HIT,MISS) + (MISS,HIT) + (MISS,MISS)}
```

The Fastly-Honeycomb integration treats the user-to-edge and edge-to-shield requests as separate, so to compute the ratio in Honeycomb, you need to deduplicate those requests.

```
# this counts all requests served from Edge
# aka (HIT,HIT) + (MISS,HIT)
count (where cache=HIT and origin like "Shield")

# this counts all requests served from Shield
# aka (HIT,MISS)
count (where cache=HIT and origin like "Cloudflare")

# this counts all requests served from Origin
# aka (MISS,MISS)
count (where cache=MISS and origin like "Cloudflare")
```
Thus
```math
CHR = \frac{\text{count (where cache=HIT)}}{\text{count (where cache=HIT or origin like "Cloudflare")}}
```

or, equivalently,
```math
CHR = 1 - \frac{\text{count (where cache=MISS and origin like "Cloudflare")}}{\text{count (where not (cache=MISS and origin like "Shield"))}}
```

**Reply by gerhard** [reply]
Since we implemented that `cache-control` last week in R2, this is how our `IAD` shield `GET /master/feed` request caching changed: https://ui.honeycomb.io/changelog/datasets/fastly/result/57f88pV53K1

<img width="1704" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/e48866c0-7564-4b87-bd71-c1caac985dc7">

Since then, we had `435` shield cache hits & `18` misses, meaning `96% cache ratio`.

Before, we had `100` shield cache misses & `81` hits, or `45% cache ratio`.

If we look at all `GET /*/feed` requests, cache misses dropped by `7x` - `9k` vs `69k`: https://ui.honeycomb.io/changelog/datasets/fastly/result/c93uNo49nna
<img width="1712" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/9aca5cc4-676f-4ae2-9d68-20361206810a">

As for the response time improvement, P99 for feed cache hits went from `2.8s` to `1s`: https://ui.honeycomb.io/changelog/datasets/fastly/result/meSXvSDkTcN
<img width="1703" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/745b1f6f-a8d3-493d-bf10-c52fc92ad973">

Since we started serving pre-computed feeds from R2, p99 responses for feed misses got worse, from `6.7s` to `7.2s`: https://ui.honeycomb.io/changelog/datasets/fastly/result/meSXvSDkTcN
<img width="1703" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/2d90cc8d-c246-4762-bd9e-1220e6969825">

It's not bad news though! Tail latency improved though - `121s` vs `528s`: https://ui.honeycomb.io/changelog/datasets/fastly/result/DxFktQBbPw9
![image](https://github.com/thechangelog/changelog.com/assets/3342/7407a2b6-441c-4d24-b583-c324731259a7)

---
Question for @jerodsanto: should `GET /feed` be served by the app, or R2? https://ui.honeycomb.io/changelog/datasets/fastly/result/zbprxo8Re4B

Question for @jamesarosen: anything that I wrote above which you would like us to dig into?

**Comment by gerhard** [comment]
https://twitter.com/gerhardlazu/status/1704438524705915352

<img width="978" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/848006c9-7411-49ef-b47b-04f7d00c4581">
